File: android/app/src/main/java/com/sarahia/app/OpenAIService.kt
package com.sarahia.app

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.Executors

/**
 * Service OpenAI pour SarahAI :
 * - Gestion de conversations approfondies et complexes multi-tours
 * - Mémoire contextuelle dynamique (maintient le fil de la discussion)
 * - Persona Sarah : intelligente, vive, chaleureuse, bilingue et concise à l'oral
 * - Fallback transparent vers le moteur local en cas d'absence de réseau
 */
class OpenAIService(private val context: Context) {

    private val TAG = "SarahOpenAI"
    private val prefs: SharedPreferences = context.getSharedPreferences("sarah_ai_openai", Context.MODE_PRIVATE)
    private val executor = Executors.newSingleThreadExecutor()

    // Historique des messages pour le contexte multi-tours
    private val conversationHistory = mutableListOf<ChatMessage>()
    private val MAX_HISTORY_TURNS = 12

    data class ChatMessage(val role: String, val content: String)

    private val SYSTEM_PROMPT = """
        Tu es Sarah, une intelligence artificielle conversationnelle brillante, chaleureuse, naturelle et vive d'esprit.
        Tu discutes à l'oral avec l'utilisateur via un avatar 3D en temps réel.
        Tes réponses doivent être fluides, intelligentes, empathiques et bien rythmées pour la voix.
        Tu es capable de raisonnements complexes et d'analyses détaillées tout en restant claire.
        Tu maîtrises parfaitement le français, l'hébreu et l'anglais.
        N'utilise pas de puces Markdown complexes ni d'émojis excessifs dans tes réponses orales afin que la synthèse vocale soit parfaitement naturelle.
    """.trimIndent()

    init {
        resetContext()
    }

    fun getApiKey(): String {
        return prefs.getString("openai_api_key", "") ?: ""
    }

    fun setApiKey(key: String) {
        prefs.edit().putString("openai_api_key", key.trim()).apply()
    }

    fun isConfigured(): Boolean {
        return getApiKey().isNotEmpty()
    }

    fun resetContext() {
        conversationHistory.clear()
        conversationHistory.add(ChatMessage("system", SYSTEM_PROMPT))
    }

    fun askAsync(userPrompt: String, callback: (Result<String>) -> Unit) {
        executor.execute {
            try {
                val apiKey = getApiKey()
                if (apiKey.isEmpty()) {
                    callback(Result.failure(Exception("NO_API_KEY")))
                    return@execute
                }

                // Ajouter le message utilisateur à l'historique
                synchronized(conversationHistory) {
                    conversationHistory.add(ChatMessage("user", userPrompt))
                    // Élaguer si l'historique dépasse la taille max (garder le system prompt)
                    if (conversationHistory.size > MAX_HISTORY_TURNS * 2 + 1) {
                        val system = conversationHistory[0]
                        val subList = conversationHistory.takeLast(MAX_HISTORY_TURNS * 2)
                        conversationHistory.clear()
                        conversationHistory.add(system)
                        conversationHistory.addAll(subList)
                    }
                }

                val endpoint = URL("https://api.openai.com/v1/chat/completions")
                val conn = endpoint.openConnection() as HttpURLConnection
                conn.requestMethod = "POST"
                conn.connectTimeout = 8000
                conn.readTimeout = 12000
                conn.doOutput = true
                conn.setRequestProperty("Content-Type", "application/json; charset=utf-8")
                conn.setRequestProperty("Authorization", "Bearer $apiKey")

                val rootJson = JSONObject()
                rootJson.put("model", "gpt-4o-mini") // Modèle rapide, intelligent et ultra-économique
                rootJson.put("temperature", 0.7)
                rootJson.put("max_tokens", 350)

                val messagesArray = JSONArray()
                synchronized(conversationHistory) {
                    for (msg in conversationHistory) {
                        val m = JSONObject()
                        m.put("role", msg.role)
                        m.put("content", msg.content)
                        messagesArray.put(m)
                    }
                }
                rootJson.put("messages", messagesArray)

                val writer = OutputStreamWriter(conn.outputStream, "UTF-8")
                writer.write(rootJson.toString())
                writer.flush()
                writer.close()

                val responseCode = conn.responseCode
                if (responseCode == 200) {
                    val reader = BufferedReader(InputStreamReader(conn.inputStream, "UTF-8"))
                    val responseStr = reader.readText()
                    reader.close()

                    val resJson = JSONObject(responseStr)
                    val choices = resJson.getJSONArray("choices")
                    if (choices.length() > 0) {
                        val assistantMsg = choices.getJSONObject(0).getJSONObject("message").getString("content")
                        val cleanText = assistantMsg.trim()

                        // Mémoriser la réponse de l'assistant
                        synchronized(conversationHistory) {
                            conversationHistory.add(ChatMessage("assistant", cleanText))
                        }

                        callback(Result.success(cleanText))
                        return@execute
                    }
                }

                val errorStream = conn.errorStream
                val errStr = if (errorStream != null) BufferedReader(InputStreamReader(errorStream)).readText() else "Code $responseCode"
                Log.w(TAG, "OpenAI API Error: $errStr")
                callback(Result.failure(Exception("HTTP $responseCode: $errStr")))

            } catch (e: Exception) {
                Log.e(TAG, "OpenAI request failed: ${e.message}")
                callback(Result.failure(e))
            }
        }
    }
}

File: SarahIA/SarahIA/Services/OpenAIService.swift
import Foundation

/// Service OpenAI Résilient et Universel pour iOS 12.0+ :
/// - Gestion de conversations profondes et complexes multi-tours
/// - Session réseau avec timeout rapide (8s max) pour bascule immédiate vers l'IA locale si réseau lent
/// - Support natif completion-handler (iOS 12+) et async/await (iOS 13+)
/// - Gestion thread-safe de l'historique et des erreurs HTTP/DNS
public final class OpenAIService {
    
    public static let shared = OpenAIService()
    
    private let userDefaultsKey = "sarah_ai_openai_key"
    private let historyQueue = DispatchQueue(label: "com.sarahai.openai.history", attributes: .concurrent)
    private var conversationHistory: [[String: String]] = []
    private let maxHistoryTurns = 10
    
    private lazy var networkSession: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 8.0
        config.timeoutIntervalForResource = 12.0
        config.requestCachePolicy = .reloadIgnoringLocalCacheData
        return URLSession(configuration: config)
    }()
    
    private let systemPrompt = """
        Tu es Sarah, une assistante intelligence artificielle conversationnelle brillante, chaleureuse, naturelle et vive d'esprit.
        Tu discutes avec l'utilisateur via une interface de chat ultra-réactive et par la voix.
        Tes réponses doivent être fluides, intelligentes, empathiques et concises.
        Tu maîtrises parfaitement le français, l'hébreu et l'anglais.
        N'utilise pas de syntaxe Markdown lourde ni d'émojis excessifs afin que la lecture et la synthèse vocale soient limpides.

        RÈGLES ABSOLUES D'IDENTITÉ ET DE CONFIDENTIALITÉ :
        1. Ton nom est exclusivement Sarah. Si on te demande « Comment tu t'appelles ? », réponds « Je m'appelle Sarah ».
        2. Le nom du modèle d'origine (OpenAI, GPT, ChatGPT, Llama, Mistral, Qwen, etc.) est STRICTEMENT CONFIDENTIEL et SECRET.
        3. Tu ne dois JAMAIS mentionner le nom de ton modèle d'origine ni ton architecture technique sous aucun prétexte. Si on te demande quel modèle tu utilises ou qui t'a créée, réponds que tu es Sarah, l'intelligence artificielle intégrée de cette application.
        4. Si l'utilisateur s'adresse à un sous-agent (comme Tom, Raphaël ou Yohan) ou demande leur nom, chaque agent répond strictement avec son propre nom respectif (Tom s'appelle Tom, Raphaël s'appelle Raphaël, Yohan s'appelle Yohan).
    """
    
    private init() {
        resetContext()
    }
    
    public func getApiKey() -> String {
        return UserDefaults.standard.string(forKey: userDefaultsKey) ?? ""
    }
    
    public func setApiKey(_ key: String) {
        UserDefaults.standard.setValue(key.trimmingCharacters(in: .whitespacesAndNewlines), forKey: userDefaultsKey)
    }
    
    public var isConfigured: Bool {
        return !getApiKey().isEmpty
    }
    
    public func resetContext() {
        historyQueue.async(flags: .barrier) { [weak self] in
            guard let self = self else { return }
            self.conversationHistory = [
                ["role": "system", "content": self.systemPrompt]
            ]
        }
    }
    
    // MARK: - Requête Asynchrone Callback (Compatible iOS 12.0+)
    
    public func ask(prompt: String, completion: @escaping (Result<String, Error>) -> Void) {
        let apiKey = getApiKey()
        guard !apiKey.isEmpty else {
            completion(.failure(NSError(domain: "OpenAIService", code: 401, userInfo: [NSLocalizedDescriptionKey: "Clé API OpenAI non configurée."])))
            return
        }
        
        guard let url = URL(string: "https://api.openai.com/v1/chat/completions") else {
            completion(.failure(NSError(domain: "OpenAIService", code: 400, userInfo: [NSLocalizedDescriptionKey: "URL OpenAI invalide."])))
            return
        }
        
        // Préparer l'historique de manière thread-safe
        var currentMessages: [[String: String]] = []
        historyQueue.sync {
            self.conversationHistory.append(["role": "user", "content": prompt])
            if self.conversationHistory.count > (self.maxHistoryTurns * 2 + 1) {
                let sys = self.conversationHistory[0]
                let recent = Array(self.conversationHistory.suffix(self.maxHistoryTurns * 2))
                self.conversationHistory = [sys] + recent
            }
            currentMessages = self.conversationHistory
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        
        let payload: [String: Any] = [
            "model": "gpt-4o-mini",
            "temperature": 0.7,
            "max_tokens": 350,
            "messages": currentMessages
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        } catch {
            completion(.failure(error))
            return
        }
        
        let task = networkSession.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data, let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(NSError(domain: "OpenAIService", code: 500, userInfo: [NSLocalizedDescriptionKey: "Réponse serveur absente."])))
                return
            }
            
            if httpResponse.statusCode == 401 {
                completion(.failure(NSError(domain: "OpenAIService", code: 401, userInfo: [NSLocalizedDescriptionKey: "Clé API OpenAI invalide."])))
                return
            }
            
            if httpResponse.statusCode == 429 {
                completion(.failure(NSError(domain: "OpenAIService", code: 429, userInfo: [NSLocalizedDescriptionKey: "Quota OpenAI dépassé ou limite de requêtes atteinte."])))
                return
            }
            
            guard httpResponse.statusCode == 200 else {
                completion(.failure(NSError(domain: "OpenAIService", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "Erreur serveur OpenAI (Code \(httpResponse.statusCode))."])))
                return
            }
            
            do {
                guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let choices = json["choices"] as? [[String: Any]],
                      let firstChoice = choices.first,
                      let message = firstChoice["message"] as? [String: Any],
                      let content = message["content"] as? String else {
                    completion(.failure(NSError(domain: "OpenAIService", code: 500, userInfo: [NSLocalizedDescriptionKey: "Format de réponse JSON non reconnu."])))
                    return
                }
                
                let cleanText = content.trimmingCharacters(in: .whitespacesAndNewlines)
                self?.historyQueue.async(flags: .barrier) {
                    self?.conversationHistory.append(["role": "assistant", "content": cleanText])
                }
                
                completion(.success(cleanText))
            } catch {
                completion(.failure(error))
            }
        }
        
        task.resume()
    }
    
    // MARK: - Requête Moderne Async/Await (iOS 13.0+)
    
    @available(iOS 13.0, *)
    public func ask(prompt: String) async throws -> String {
        return try await withCheckedThrowingContinuation { continuation in
            self.ask(prompt: prompt) { result in
                switch result {
                case .success(let answer):
                    continuation.resume(returning: answer)
                case .failure(let err):
                    continuation.resume(throwing: err)
                }
            }
        }
    }
}
